### PR TITLE
Add types

### PR DESCRIPTION
**Pull request by Jack-Works**
Create `index.d.ts`.

### CODE REVIEW

**Comment by rekmarks** [comment]
Thanks for making this, @Jack-Works! We've started to implement some types for this internally, but never finished the job. We'll be reviewing this in the coming days 👍

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import { EventEmitter } from 'events';
+import { Duplex } from 'stream';
+import { AbstractProvider } from 'web3-core';
```
Comment (danfinlay) [inline index.d.ts]: I see web3-core is LGPL, which could taint the license of our whole project, so we might want to just write this fairly simple part under an even more permissive license.

**Review by danfinlay — DISMISSED** [review]
Overall a great addition, but we should do it without license changes.

File: index.d.ts (hunk 1 above)
Comment (Jack-Works) [reply index.d.ts]: AbstractProvider is a simple interface I can add it without add web3-core as a dependency. I will do the change later

**Comment by rekmarks** [comment]
@Jack-Works for your reference, @brad-decker (on our team) previously sketched the following types for this package. I'm not suggesting you have to adapt any of this over what you've done, but in case it's useful: https://gist.github.com/rekmarks/1d249cb9d805f8b8ad89467ae961517b

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+     * @param options An options bag
+     */
+    constructor(connectionStream: Duplex, options?: MetamaskInpageProviderOptions);
+    readonly isMetaMask: true;
+    readonly selectedAddress: string | null;
+    // Help wanted. Is it a string?
+    readonly networkVersion: undefined;
+    // Help wanted. Is it a string?
+    readonly chainId: undefined;
+    get publicConfigStore(): any;
```
Comment (rekmarks) [inline index.d.ts]: We're planning to remove the `publicConfigStore` property, so I'm not sure if we should include it or not 🤔

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+     * @param connectionStream A Node.js duplex stream
+     * @param options An options bag
+     */
+    constructor(connectionStream: Duplex, options?: MetamaskInpageProviderOptions);
+    readonly isMetaMask: true;
+    readonly selectedAddress: string | null;
+    // Help wanted. Is it a string?
+    readonly networkVersion: undefined;
+    // Help wanted. Is it a string?
+    readonly chainId: undefined;
```
Comment (rekmarks) [inline index.d.ts]: They should be `string | null`

ref: #93

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import { EventEmitter } from 'events';
+import { Duplex } from 'stream';
+import { AbstractProvider } from 'web3-core';
+
+export interface MetamaskInpageProviderOptions {
+    /**
+     * The logging API to use.
+     * @default console
+     */
+    logger?: Console;
```
Comment (rekmarks) [inline index.d.ts]: Could we extend this to allow other loggers? For reference, we should probably turn this into a type: https://github.com/MetaMask/inpage-provider/blob/master/src/MetamaskInpageProvider.js/#L696

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    readonly isMetaMask: true;
+    readonly selectedAddress: string | null;
+    // Help wanted. Is it a string?
+    readonly networkVersion: undefined;
+    // Help wanted. Is it a string?
+    readonly chainId: undefined;
+    get publicConfigStore(): any;
+    isConnected(): boolean;
+    /** Submits an RPC request to MetaMask for the given method, with the given params. Resolves with the result of the method call, or rejects on error. */
+    request(jsonRPCPayload: object): Promise<unknown>;
```
Comment (rekmarks) [inline index.d.ts]: The `request` parameter should follow the `RequestArguments` type defined in EIP-1193: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md#request

I also think it's in `web3-core`.

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    sendAsync(jsonRPCPayload: object, callback: (err: Error | undefined, result: undefined) => void): void;
```
Comment (rekmarks) [inline index.d.ts]: The `jsonRPCPayload` should be typed like it is in `web3-core`.

File: index.d.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+export function setGlobalProvider(providerInstance: MetamaskInpageProvider): void
+export interface JsonRpcPayload {
+    jsonrpc: string
+    method: string
+    params: unknown[]
+    id?: string | number
+}
+export interface JsonRpcResponse {
+    jsonrpc: string
+    id: number
```
Comment (rekmarks) [inline index.d.ts]:
```suggestion
    id: string | number
```

File: index.d.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+/**
+ * Sets the given provider instance as window.ethereum and dispatches the 'ethereum#initialized' event on window.
+ *
+ * @param providerInstance - The provider instance.
+ */
+export function setGlobalProvider(providerInstance: MetamaskInpageProvider): void
+export interface JsonRpcPayload {
+    jsonrpc: string
+    method: string
+    params: unknown[]
```
Comment (rekmarks) [inline index.d.ts]: `params` are not always required.
```suggestion
    params?: unknown[]
```

File: index.d.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    id?: string | number
+}
+export interface JsonRpcResponse {
+    jsonrpc: string
+    id: number
+    result?: unknown
+    error?: string
```
Comment (rekmarks) [inline index.d.ts]: All RPC errors (from `sendAsync`, `request`, `send`) follow this interface, without the `stack`: https://github.com/MetaMask/eth-rpc-errors/blob/master/%40types/index.d.ts#L2-L7

**Review by brad-decker — DISMISSED** [review]
LGTM, good work!

File: index.d.ts [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+}
+export interface JsonRpcResponse {
+    jsonrpc: string
+    id: string | number
+    result?: unknown
+    error?: {
+        code: number
+        message: string
+        data?: unknown
+        stack?: any
```
Comment (rekmarks) [inline index.d.ts]: The `stack` will never be present here in practice:
```suggestion
```

File: index.d.ts [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+     */
+    constructor(connectionStream: Duplex, options?: MetamaskInpageProviderOptions)
+    readonly isMetaMask: true
+    readonly selectedAddress: string | null
+    readonly networkVersion: string | null
+    readonly chainId: string | undefined
+    isConnected(): boolean
+    sendAsync(payload: JsonRpcPayload, callback: (error: Error | null, result?: JsonRpcResponse) => void): void
+    /** Submits an RPC request to MetaMask for the given method, with the given params. Resolves with the result of the method call, or rejects on error. */
+    request(args: JsonRpcPayload): Promise<any>
```
Comment (rekmarks) [inline index.d.ts]: We should prefer `unknown`:
```suggestion
    request(args: JsonRpcPayload): Promise<unknown>
```

**Review by rekmarks — CHANGES_REQUESTED** [review]
Just a couple more pieces of minor changes, then we should be good to go!

**Review by rekmarks — APPROVED** [review]
LGTM!

Great work @Jack-Works. Many thanks!

**Pull request merged**